pkg/api/persistentvolumeclaim/util_test.go

```diff
@@ -297,3 +303,104 @@ func TestPVCDataSourceSpecFilter(t *testing.T) {
+	makeDataSource := func(apiGroup, kind, name string) *core.TypedLocalObjectReference {
+		return &core.TypedLocalObjectReference{
+			APIGroup: &apiGroup,
+			Kind:     kind,
+			Name:     name,
+		}
+	}
+
+	volumeDataSource := makeDataSource("", "PersistentVolumeClaim", "my-vol")
+	snapshotDataSource := makeDataSource("snapshot.storage.k8s.io", "VolumeSnapshot", "my-snap")
+	genericDataSource := makeDataSource("generic.storage.k8s.io", "Generic", "my-foo")
+
+	var tests = map[string]struct {
+		spec            core.PersistentVolumeClaimSpec
+		volumeEnabled   bool
+		snapshotEnabled bool
+		genericEnabled  bool
+		want            *core.TypedLocalObjectReference
+	}{
+		"all disabled with empty ds": {
+			spec: core.PersistentVolumeClaimSpec{},
+			want: nil,
+		},
+		"all disabled with volume ds": {
+			spec: core.PersistentVolumeClaimSpec{DataSource: volumeDataSource},
+			want: nil,
+		},
+		"all disabled with snapshot ds": {
+			spec: core.PersistentVolumeClaimSpec{DataSource: snapshotDataSource},
+			want: nil,
+		},
+		"all disabled with generic ds": {
+			spec: core.PersistentVolumeClaimSpec{DataSource: genericDataSource},
+			want: nil,
+		},
+		"generic enabled with empty ds": {
+			spec:           core.PersistentVolumeClaimSpec{},
+			genericEnabled: true,
+			want:           nil,
+		},
+		"generic enabled with volume ds": {
+			spec:           core.PersistentVolumeClaimSpec{DataSource: volumeDataSource},
+			genericEnabled: true,
+			want:           volumeDataSource,
+		},
+		"generic enabled with snapshot ds": {
+			spec:           core.PersistentVolumeClaimSpec{DataSource: snapshotDataSource},
+			genericEnabled: true,
+			want:           snapshotDataSource,
+		},
+		"generic enabled with generic ds": {
+			spec:           core.PersistentVolumeClaimSpec{DataSource: genericDataSource},
+			genericEnabled: true,
+			want:           genericDataSource,
+		},
+		"volume enabled with volume ds": {
+			spec:          core.PersistentVolumeClaimSpec{DataSource: volumeDataSource},
+			volumeEnabled: true,
+			want:          volumeDataSource,
+		},
+		"snapshot enabled with snapshot ds": {
+			spec:            core.PersistentVolumeClaimSpec{DataSource: snapshotDataSource},
+			snapshotEnabled: true,
+			want:            snapshotDataSource,
+		},
+		"volume and generic enabled with volume ds": {
+			spec:           core.PersistentVolumeClaimSpec{DataSource: volumeDataSource},
+			volumeEnabled:  true,
+			genericEnabled: true,
+			want:           volumeDataSource,
+		},
+		"snapshot and generic enabled with snapshot ds": {
+			spec:            core.PersistentVolumeClaimSpec{DataSource: snapshotDataSource},
+			snapshotEnabled: true,
+			genericEnabled:  true,
+			want:            snapshotDataSource,
+		},
+		"volume and snapshot enabled with generic ds": {
+			spec:            core.PersistentVolumeClaimSpec{DataSource: genericDataSource},
+			volumeEnabled:   true,
+			snapshotEnabled: true,
+			want:            nil,
+		},
+	}
```

Comment:
The idea is that if the new generic data source feature gate is enabled, it doesn't matter what the other 2 feature gates are set to, because anything is allowed. In theory these 3 feature gates could be set 8 different ways, but it's not worth testing all the combinations. I focused on the worthwhile combinations here with my tests.

Reply:
> The idea is that if the new generic data source feature gate is enabled, it doesn't matter what the other 2 feature gates are set to

Agreed, but I would argue that it is worth testing all of the combinations.